File: lib/offline/bot/play_strategy.dart
import 'package:koutbh/shared/models/card.dart';
import 'package:koutbh/shared/models/game_state.dart';
import 'package:koutbh/shared/logic/play_validator.dart';
import 'package:koutbh/shared/logic/trick_resolver.dart';
import 'package:koutbh/shared/logic/card_utils.dart';
import 'package:koutbh/offline/player_controller.dart';
import 'package:koutbh/offline/bot/card_tracker.dart';
import 'package:koutbh/offline/bot/game_context.dart';

class PlayStrategy {
  static PlayCardAction selectCard({
    required List<GameCard> hand,
    required List<({String playerUid, GameCard card})> trickPlays,
    required Suit? trumpSuit,
    required Suit? ledSuit,
    required int mySeat,
    String? partnerUid,
    bool isKout = false,
    bool isFirstTrick = false,
    GameContext? context,
  }) {
    final legalCards = _legalPlays(
      hand,
      ledSuit,
      trickPlays.isEmpty,
      trumpSuit: trumpSuit,
      isKout: isKout,
      isFirstTrick: isFirstTrick,
    );

    if (legalCards.length == 1) {
      return PlayCardAction(legalCards.first);
    }

    if (trickPlays.isEmpty) {
      return PlayCardAction(
        _selectLead(legalCards, trumpSuit, context: context, hand: hand),
      );
    }

    return PlayCardAction(
      _selectFollow(
        legalCards: legalCards,
        hand: hand,
        trickPlays: trickPlays,
        trumpSuit: trumpSuit,
        ledSuit: ledSuit,
        partnerUid: partnerUid,
        context: context,
      ),
    );
  }

  static List<GameCard> _legalPlays(
    List<GameCard> hand,
    Suit? ledSuit,
    bool isLead, {
    Suit? trumpSuit,
    bool isKout = false,
    bool isFirstTrick = false,
  }) {
    final legal = PlayValidator.playableForCurrentTrick(
      hand: hand,
      trickHasNoPlaysYet: isLead,
      ledSuit: ledSuit,
      trumpSuit: trumpSuit,
      bidIsKout: isKout,
      noTricksCompletedYet: isFirstTrick,
    ).toList();

    if (isLead) {
      final nonJoker = legal.where((c) => !c.isJoker).toList();
      if (nonJoker.isNotEmpty) return nonJoker;
    }

    return legal;
  }

  static bool _isLowLead(GameCard card) {
    if (card.isJoker || card.rank == null) return false;
    return card.rank!.value <= Rank.ten.value;
  }

  static bool _defenderDumpMode(GameContext ctx) {
    if (ctx.isBiddingTeam) return false;
    if (ctx.bidderSeat == null) return false;
    final bidderTeam = teamForSeat(ctx.bidderSeat!);
    final won = ctx.trickCounts[bidderTeam] ?? 0;
    return won >= (ctx.currentBid?.value ?? 5);
  }

  static bool _defenderBidAlreadyLost(GameContext ctx) {
    if (ctx.isBiddingTeam) return false;
    if (ctx.bidderSeat == null) return false;
    final bidderTeam = teamForSeat(ctx.bidderSeat!);
    final won = ctx.trickCounts[bidderTeam] ?? 0;
    final need = (ctx.currentBid?.value ?? 5) - won;
    final remaining = 8 - ctx.tricksPlayed;
    return need > remaining;
  }

  static GameCard _selectLead(
    List<GameCard> legalCards,
    Suit? trumpSuit, {
    GameContext? context,
    List<GameCard>? hand,
  }) {
    final fullHand = hand ?? legalCards;

    if (context?.tracker != null) {
      final masters = legalCards
          .where(
            (c) =>
                !c.isJoker && context!.tracker!.isHighestRemaining(c, fullHand),
          )
          .toList();
      if (masters.isNotEmpty) {
        masters.sort((a, b) {
          final aT = a.suit == trumpSuit;
          final bT = b.suit == trumpSuit;
          if (aT != bT) return aT ? 1 : -1;
          return b.rank!.value.compareTo(a.rank!.value);
        });
        return masters.first;
      }
    }

    final aces = legalCards
        .where(
          (c) =>
              !c.isJoker &&
              c.rank == Rank.ace &&
              (trumpSuit == null || c.suit != trumpSuit),
        )
        .toList();

    if (aces.isNotEmpty) {
      final aceWithKing = aces.where(
        (a) => legalCards.any(
          (c) => !c.isJoker && c.suit == a.suit && c.rank == Rank.king,
        ),
      );
      if (aceWithKing.isNotEmpty) return aceWithKing.first;

      final singletonAce = aces.where(
        (a) =>
            legalCards.where((c) => !c.isJoker && c.suit == a.suit).length == 1,
      );
      if (singletonAce.isNotEmpty) return singletonAce.first;

      return aces.first;
    }

    if (context != null && context.isBiddingTeam && trumpSuit != null) {
      final myTrumps = legalCards
          .where((c) => !c.isJoker && c.suit == trumpSuit)
          .toList();
      if (myTrumps.length >= 3) {
        myTrumps.sort((a, b) => b.rank!.value.compareTo(a.rank!.value));
        return myTrumps.first;
      }
    }

    if (context?.tracker != null) {
      final partnerSeat = context!.partnerSeat;
      final partnerVoids = context.tracker!.knownVoids[partnerSeat] ?? {};
      for (final voidSuit in partnerVoids) {
        if (voidSuit == trumpSuit) continue;
        final suitCards = legalCards
            .where((c) => !c.isJoker && c.suit == voidSuit)
            .toList();
        if (suitCards.isNotEmpty) {
          suitCards.sort((a, b) => a.rank!.value.compareTo(b.rank!.value));
          return suitCards.first;
        }
      }
    }

    if (context != null && !context.isBiddingTeam) {
      final nonTrumpSingles = legalCards.where((c) {
        if (c.isJoker) return false;
        if (c.suit == trumpSuit) return false;
        return legalCards.where((o) => !o.isJoker && o.suit == c.suit).length ==
            1;
      }).toList();
      if (nonTrumpSingles.isNotEmpty) return nonTrumpSingles.first;
    }

    final nonTrump = trumpSuit != null
        ? legalCards.where((c) => !c.isJoker && c.suit != trumpSuit).toList()
        : legalCards.where((c) => !c.isJoker).toList();

    if (nonTrump.isNotEmpty) {
      final counts = countBySuit(nonTrump);
      final suitGroups = <Suit, List<GameCard>>{};
      for (final c in nonTrump) {
        suitGroups.putIfAbsent(c.suit!, () => []).add(c);
      }

      final sortedSuits = suitGroups.entries.toList()
        ..sort((a, b) {
          final lenCmp = counts[b.key]!.compareTo(counts[a.key]!);
          if (lenCmp != 0) return lenCmp;
          final aMax = a.value
              .map((c) => c.rank!.value)
              .reduce((a, b) => a > b ? a : b);
          final bMax = b.value
              .map((c) => c.rank!.value)
              .reduce((a, b) => a > b ? a : b);
          return bMax.compareTo(aMax);
        });

      final bestSuitCards = sortedSuits.first.value;
      bestSuitCards.sort((a, b) => a.rank!.value.compareTo(b.rank!.value));
      return bestSuitCards.first;
    }

    final sorted = legalCards.where((c) => !c.isJoker).toList()
      ..sort((a, b) => b.rank!.value.compareTo(a.rank!.value));
    return sorted.isNotEmpty ? sorted.first : legalCards.first;
  }

  static GameCard _selectFollow({
    required List<GameCard> legalCards,
    required List<GameCard> hand,
    required List<({String playerUid, GameCard card})> trickPlays,
    required Suit? trumpSuit,
    required Suit? ledSuit,
    String? partnerUid,
    GameContext? context,
  }) {
    final winningPlay = _winningPlay(trickPlays, trumpSuit, ledSuit);
    final partnerWinning =
        partnerUid != null && winningPlay?.playerUid == partnerUid;
    final hasJoker = legalCards.any((c) => c.isJoker);
    final myPosition = trickPlays.length;

    final followingSuit =
        ledSuit != null &&
        legalCards.every((c) => !c.isJoker && c.suit == ledSuit);

    if (context != null) {
      if (context.isBiddingTeam && context.tricksNeededForBid <= 0) {
        return _strategicDump(legalCards, hand, trumpSuit);
      }
      if (_defenderDumpMode(context) || _defenderBidAlreadyLost(context)) {
        return _strategicDump(legalCards, hand, trumpSuit);
      }
      if (!context.isBiddingTeam && context.tricksNeededForBid == 1) {
        if (hasJoker) return legalCards.firstWhere((c) => c.isJoker);
        if (trumpSuit != null) {
          final trumpCards = legalCards
              .where((c) => !c.isJoker && c.suit == trumpSuit)
              .toList();
          final winningTrumps = _cardsBeating(
            trumpCards,
            trickPlays,
            trumpSuit,
            ledSuit,
          );
          if (winningTrumps.isNotEmpty) return _lowest(winningTrumps);
        }
      }
    }

    if (partnerUid != null &&
        trickPlays.isNotEmpty &&
        trickPlays.first.playerUid == partnerUid &&
        _isLowLead(trickPlays.first.card) &&
        (myPosition == 1 || myPosition == 2)) {
      final w = _winningPlay(trickPlays, trumpSuit, ledSuit);
      if (w != null && w.playerUid == partnerUid) {
        return _lowest(legalCards);
      }
    }

    if (context?.isForcedBid == true) {
      if (followingSuit) {
        final aces = legalCards.where((c) => c.rank == Rank.ace).toList();
        if (aces.isNotEmpty) return aces.first;
        return _lowest(legalCards);
      }
      return _strategicDump(legalCards, hand, trumpSuit);
    }

    if (hand.length <= 2 &&
        hasJoker &&
        hand.where((c) => !c.isJoker).length <= 1) {
      return legalCards.firstWhere((c) => c.isJoker);
    }

    if (followingSuit) {
      final urgency = context?.roundControlUrgency ?? 0;
      final canBeat = _cardsBeating(legalCards, trickPlays, trumpSuit, ledSuit);
      final shouldProtectPartner =
          partnerWinning && myPosition <= 2 && urgency < 0.8;
      if (shouldProtectPartner) {
        return _lowest(legalCards);
      }
      final mustSecureNow =
          partnerWinning && urgency >= 0.8 && canBeat.isNotEmpty;
      if (mustSecureNow) {
        return _lowest(canBeat);
      }

      // Last to play: if partner already has the trick, do not overtake.
      if (partnerWinning && myPosition == 3) {
        return _lowest(legalCards);
      }
      if (myPosition == 3) {
        final winners = _cardsBeating(
          legalCards,
          trickPlays,
          trumpSuit,
          ledSuit,
        );
        return _lowest(winners.isNotEmpty ? winners : legalCards);
      } else {
        final winners = _cardsBeating(
          legalCards,
          trickPlays,
          trumpSuit,
          ledSuit,
        );
        if (winners.isNotEmpty) {
          return _lowest(winners);
        }
        return _lowest(legalCards);
      }
    }

    if (partnerWinning) {
      if (hasJoker) {
        final nonJoker = legalCards.where((c) => !c.isJoker).toList();
        if (_jokerPoisonRisk(nonJoker, hand, context?.tracker)) {
          return legalCards.firstWhere((c) => c.isJoker);
        }
      }
      return _strategicDump(legalCards, hand, trumpSuit);
    }

    if (hasJoker) {
      final nonJoker = legalCards.where((c) => !c.isJoker).toList();

      if (_cannotWinWithoutJoker(nonJoker, trickPlays, trumpSuit, ledSuit) &&
          !partnerWinning) {
        return legalCards.firstWhere((c) => c.isJoker);
      }

      if (_jokerPoisonRisk(nonJoker, hand, context?.tracker)) {
        return legalCards.firstWhere((c) => c.isJoker);
      }

      double urgency = 0.0;

      if (context != null) {
        final needed = context.isBiddingTeam
            ? context.tricksNeededForBid
            : (8 - (context.currentBid?.value ?? 5) + 1) -
                  context.opponentTricks;
        if (needed == 1) urgency += 0.5;
      }

      final opponentTrumped =
          trumpSuit != null &&
          trickPlays.any((p) => !p.card.isJoker && p.card.suit == trumpSuit);
      if (opponentTrumped) urgency += 0.3;

      if (hand.length <= 3) urgency += 0.3;

      if (partnerWinning) urgency -= 0.8;

      // Threshold inlined; PlayStrategy will be rewritten in Task 8.
      if (urgency >= 0.08) {
        return legalCards.firstWhere((c) => c.isJoker);
      }
    }

    if (context?.tracker != null && trumpSuit != null) {
      final trumpsOut = context!.tracker!.trumpsRemaining(trumpSuit, hand);
      final myTrumps = legalCards
          .where((c) => !c.isJoker && c.suit == trumpSuit)
          .toList();
      if (trumpsOut <= 1 && myTrumps.length == 1) {
        final trickHasHonor = _trickHasAceOrKing(trickPlays);
        if (!trickHasHonor && (context.tricksNeededForBid) > 1) {
          return _strategicDump(legalCards, hand, trumpSuit);
        }
      }
    }

    if (trumpSuit != null) {
      final trumpCards = legalCards
          .where((c) => !c.isJoker && c.suit == trumpSuit)
          .toList();
      if (trumpCards.isNotEmpty) {
        final winningTrumps = _cardsBeating(
          trumpCards,
          trickPlays,
          trumpSuit,
          ledSuit,
        );
        if (winningTrumps.isNotEmpty) return _lowest(winningTrumps);
        return _strategicDump(legalCards, hand, trumpSuit);
      }
    }

    return _strategicDump(legalCards, hand, trumpSuit);
  }

  static GameCard _strategicDump(
    List<GameCard> legalCards,
    List<GameCard> hand,
    Suit? trumpSuit,
  ) {
    final dumpable = legalCards.where((c) => !c.isJoker).toList();
    if (dumpable.isEmpty) return legalCards.first;

    final singletons = dumpable.where((c) {
      final suitCount = hand
          .where((h) => !h.isJoker && h.suit == c.suit)
          .length;
      return suitCount == 1 && c.suit != trumpSuit;
    }).toList();
    if (singletons.isNotEmpty) {
      singletons.sort((a, b) => a.rank!.value.compareTo(b.rank!.value));
      return singletons.first;
    }

    final safeToBreak = dumpable.where((c) {
      if (c.suit == trumpSuit) return false;
      if (c.rank == Rank.king &&
          hand.any((h) => h.suit == c.suit && h.rank == Rank.ace)) {
        return false;
      }
      if (c.rank == Rank.queen &&
          hand.any((h) => h.suit == c.suit && h.rank == Rank.king)) {
        return false;
      }
      return true;
    }).toList();

    if (safeToBreak.isNotEmpty) {
      safeToBreak.sort((a, b) => a.rank!.value.compareTo(b.rank!.value));
      return safeToBreak.first;
    }

    final nonTrump = dumpable.where((c) => c.suit != trumpSuit).toList();
    if (nonTrump.isNotEmpty) {
      nonTrump.sort((a, b) => a.rank!.value.compareTo(b.rank!.value));
      return nonTrump.first;
    }
    dumpable.sort((a, b) => a.rank!.value.compareTo(b.rank!.value));
    return dumpable.first;
  }

  static GameCard _lowest(List<GameCard> cards) {
    final nonJoker = cards.where((c) => !c.isJoker).toList();
    if (nonJoker.isEmpty) return cards.first;
    nonJoker.sort((a, b) => a.rank!.value.compareTo(b.rank!.value));
    return nonJoker.first;
  }

  static ({String playerUid, GameCard card})? _winningPlay(
    List<({String playerUid, GameCard card})> plays,
    Suit? trumpSuit,
    Suit? ledSuit,
  ) {
    if (plays.isEmpty) return null;

    for (final play in plays) {
      if (play.card.isJoker) return play;
    }

    if (ledSuit == null && plays.isNotEmpty && !plays.first.card.isJoker) {
      ledSuit = plays.first.card.suit;
    }

    var bestPlay = plays.first;
    for (int i = 1; i < plays.length; i++) {
      if (TrickResolver.beats(
        plays[i].card,
        bestPlay.card,
        trumpSuit,
        ledSuit,
      )) {
        bestPlay = plays[i];
      }
    }
    return bestPlay;
  }

  static bool _jokerPoisonRisk(
    List<GameCard> nonJokerLegal,
    List<GameCard> hand,
    CardTracker? tracker,
  ) {
    if (nonJokerLegal.isEmpty) return true;
    if (nonJokerLegal.length <= 1) return true;
    if (tracker != null && nonJokerLegal.length <= 2) {
      return nonJokerLegal.any((c) => !tracker.isSuitExhausted(c.suit!, hand));
    }
    return false;
  }

  static bool _trickHasAceOrKing(
    List<({String playerUid, GameCard card})> trickPlays,
  ) {
    return trickPlays.any(
      (p) =>
          !p.card.isJoker &&
          (p.card.rank == Rank.ace || p.card.rank == Rank.king),
    );
  }

  static bool _cannotWinWithoutJoker(
    List<GameCard> nonJokerLegal,
    List<({String playerUid, GameCard card})> trickPlays,
    Suit? trumpSuit,
    Suit? ledSuit,
  ) {
    if (nonJokerLegal.isEmpty) return true;
    final winners = _cardsBeating(
      nonJokerLegal,
      trickPlays,
      trumpSuit,
      ledSuit,
    );
    return winners.isEmpty;
  }

  static List<GameCard> _cardsBeating(
    List<GameCard> candidates,
    List<({String playerUid, GameCard card})> trickPlays,
    Suit? trumpSuit,
    Suit? ledSuit,
  ) {
    final best = _winningPlay(trickPlays, trumpSuit, ledSuit);
    if (best == null) return candidates;

    return candidates
        .where((c) => TrickResolver.beats(c, best.card, trumpSuit, ledSuit))
        .toList();
  }
}

File: workers/src/game/bot/play-strategy.ts
import { decodeCard, beatsCard } from '../card';
import { RANK_VALUES, TRICKS_PER_ROUND } from '../types';
import type { SuitName, TrickPlay } from '../types';
import type { CardTracker } from './card-tracker';
import type { BotContext } from './types';

export function decidePlay(ctx: BotContext, tracker: CardTracker): string {
  const isKout = ctx.currentBid === 8;
  const isFirstTrick = ctx.trickWinners.length === 0;
  const legal = getLegalPlays(ctx.hand, ctx.currentTrick, ctx.trumpSuit, ctx.isLead, isKout, isFirstTrick);
  if (legal.length === 1) return legal[0];

  if (ctx.isLead) {
    return selectLead(legal, ctx, tracker);
  }
  return selectFollow(legal, ctx, tracker);
}

function getLegalPlays(
  hand: string[], trick: TrickPlay[], trumpSuit: SuitName | undefined, isLead: boolean,
  isKout: boolean = false, isFirstTrick: boolean = false,
): string[] {
  if (isLead) {
    // Kout first trick: must lead trump if you have any (matches PlayValidator).
    if (isKout && isFirstTrick && trumpSuit) {
      const trumpCards = hand.filter(c => {
        const d = decodeCard(c);
        return !d.isJoker && d.suit === trumpSuit;
      });
      if (trumpCards.length > 0) return trumpCards;
    }
    const nonJoker = hand.filter(c => c !== 'JO');
    return nonJoker.length > 0 ? nonJoker : [...hand];
  }

  if (trick.length === 0) return [...hand];

  const ledCard = decodeCard(trick[0].card);
  if (ledCard.isJoker) return [...hand];

  const ledSuit = ledCard.suit!;
  const suitCards = hand.filter(c => {
    const d = decodeCard(c);
    return !d.isJoker && d.suit === ledSuit;
  });

  if (suitCards.length > 0) {
    return hand.includes('JO') ? [...suitCards, 'JO'] : suitCards;
  }

  return [...hand];
}

function selectLead(legal: string[], ctx: BotContext, tracker: CardTracker): string {
  const { trumpSuit, hand } = ctx;

  // Master card leads — non-trump before trump.
  const masters = legal.filter(c => {
    const d = decodeCard(c);
    return !d.isJoker && tracker.isHighestRemaining(c, hand);
  });
  if (masters.length > 0) {
    // Non-trump masters first, sorted by rank descending.
    const nonTrumpMasters = masters.filter(c => decodeCard(c).suit !== trumpSuit);
    const pool = nonTrumpMasters.length > 0 ? nonTrumpMasters : masters;
    pool.sort((a, b) => rankVal(b) - rankVal(a));
    return pool[0];
  }

  // Non-trump aces — prefer A-K combo.
  const aces = legal.filter(c => {
    const d = decodeCard(c);
    return !d.isJoker && d.rank === 'ace' && d.suit !== trumpSuit;
  });
  if (aces.length > 0) {
    const aceWithKing = aces.find(a => {
      const suit = decodeCard(a).suit!;
      return hand.some(h => { const hd = decodeCard(h); return !hd.isJoker && hd.suit === suit && hd.rank === 'king'; });
    });
    return aceWithKing ?? aces[0];
  }

  // 3. Singleton voids — lead a singleton non-trump card when you have trump (play_strategy.dart).
  if (trumpSuit) {
    const hasTrump = hand.some(c => {
      const d = decodeCard(c);
      return !d.isJoker && d.suit === trumpSuit;
    });
    if (hasTrump) {
      const singletons = legal.filter(c => {
        const d = decodeCard(c);
        if (d.isJoker || d.suit === trumpSuit) return false;
        const count = hand.filter(h => {
          const hd = decodeCard(h);
          return !hd.isJoker && hd.suit === d.suit;
        }).length;
        return count === 1;
      });
      if (singletons.length > 0) {
        singletons.sort((a, b) => rankVal(a) - rankVal(b));
        return singletons[0];
      }
    }
  }

  // 4. Trump strip — bidding team with 3+ trumps: lead highest trump.
  if (ctx.isBiddingTeam && trumpSuit) {
    const myTrumps = legal.filter(c => {
      const d = decodeCard(c);
      return !d.isJoker && d.suit === trumpSuit;
    });
    if (myTrumps.length >= 3) {
      myTrumps.sort((a, b) => rankVal(b) - rankVal(a));
      return myTrumps[0];
    }
  }

  // 5. Partner void exploit.
  const partnerVoids = tracker.knownVoids.get(ctx.partnerSeat);
  if (partnerVoids && partnerVoids.size > 0) {
    for (const voidSuit of partnerVoids) {
      if (voidSuit === trumpSuit) continue;
      const suitCards = legal.filter(c => {
        const d = decodeCard(c);
        return !d.isJoker && d.suit === voidSuit;
      });
      if (suitCards.length > 0) {
        suitCards.sort((a, b) => rankVal(a) - rankVal(b));
        return suitCards[0];
      }
    }
  }

  return leadFromLongestSuit(legal, trumpSuit);
}

function selectFollow(legal: string[], ctx: BotContext, tracker: CardTracker): string {
  const { trumpSuit, currentTrick, hand, partnerSeat, players } = ctx;
  if (currentTrick.length === 0) return legal[0];

  const ledCard = decodeCard(currentTrick[0].card);
  if (ledCard.isJoker) return lowest(legal);

  const ledSuit = ledCard.suit!;
  const myPosition = currentTrick.length; // 1=2nd, 2=3rd, 3=4th
  const winning = currentWinner(currentTrick, trumpSuit, ledSuit);
  const partnerUid = players[partnerSeat];
  const partnerWinning = winning?.player === partnerUid;
  const followingSuit = legal.some(c => {
    const d = decodeCard(c);
    return !d.isJoker && d.suit === ledSuit;
  });
  const hasJoker = legal.includes('JO');
  const tricksRemaining = TRICKS_PER_ROUND - (ctx.trickWinners.length);

  // Joker countdown.
  if (tricksRemaining <= 2 && hasJoker) {
    return 'JO';
  }

  // Poison prevention.
  if (hand.length <= 2 && hasJoker && legal.includes('JO')) {
    return 'JO';
  }

  if (followingSuit) {
    return followSuit(legal, currentTrick, trumpSuit, ledSuit, partnerWinning, myPosition, hasJoker);
  }

  return voidFollow(legal, hand, currentTrick, trumpSuit, ledSuit, partnerWinning, myPosition, hasJoker, tracker);
}

function followSuit(
  legal: string[],
  trickPlays: TrickPlay[],
  trumpSuit: SuitName | undefined,
  ledSuit: SuitName,
  partnerWinning: boolean,
  myPosition: number,
  hasJoker: boolean,
): string {
  const suitCards = legal.filter(c => { const d = decodeCard(c); return !d.isJoker && d.suit === ledSuit; });
  const canBeat = cardsBeating(suitCards, trickPlays, trumpSuit, ledSuit);

  // Partner winning → play lowest.
  if (partnerWinning) return lowest(suitCards);

  if (canBeat.length > 0) {
    // Last to play: lowest winner suffices.
    if (myPosition === 3) return lowest(canBeat);
    // Otherwise: highest winner.
    return highest(canBeat);
  }

  // Cannot beat with suit cards.
  // Last to play + have Joker → play Joker to salvage the trick.
  if (myPosition === 3 && hasJoker) return 'JO';

  // Not last or no Joker → play lowest suit card.
  return lowest(suitCards);
}

function voidFollow(
  legal: string[],
  hand: string[],
  trickPlays: TrickPlay[],
  trumpSuit: SuitName | undefined,
  ledSuit: SuitName,
  partnerWinning: boolean,
  myPosition: number,
  hasJoker: boolean,
  tracker: CardTracker,
): string {
  const trumpCards = trumpSuit
    ? legal.filter(c => { const d = decodeCard(c); return !d.isJoker && d.suit === trumpSuit; })
    : [];
  const winningTrumps = cardsBeating(trumpCards, trickPlays, trumpSuit, ledSuit);

  // Partner winning safely (last to play) → dump.
  if (partnerWinning && myPosition === 3) {
    return strategicDump(legal, hand, trumpSuit);
  }

  // Partner winning but opponent still to play → consider trumping to guarantee.
  if (partnerWinning) {
    if (trumpCards.length > 0 && trumpSuit) {
      const trumpsOut = tracker.trumpsRemaining(trumpSuit, hand);
      if (trumpsOut === 0) {
        // No opponent trumps remaining — any trump guarantees.
        return lowest(trumpCards);
      }
      // Check if my highest trump beats all remaining trump.
      const remaining = tracker.remainingCards(hand);
      const remainingTrumps = remaining.filter(c => { const d = decodeCard(c); return !d.isJoker && d.suit === trumpSuit; });
      const myHighestTrump = trumpCards.reduce((best, c) => rankVal(c) > rankVal(best) ? c : best, trumpCards[0]);
      const canGuarantee = remainingTrumps.every(c => rankVal(myHighestTrump) > rankVal(c));
      if (canGuarantee) return myHighestTrump;
    }
    return strategicDump(legal, hand, trumpSuit);
  }

  // Opponent winning + have winning trump → play lowest winning trump.
  if (winningTrumps.length > 0) return lowest(winningTrumps);

  // Opponent winning + have trump but none beat current winner → play lowest trump.
  if (trumpCards.length > 0) return lowest(trumpCards);

  // Can't win with trump → try Joker if it would win.
  if (hasJoker) {
    const nonJokerCanWin = cardsBeating(
      legal.filter(c => c !== 'JO'),
      trickPlays,
      trumpSuit,
      ledSuit,
    ).length > 0;
    if (!nonJokerCanWin) return 'JO';
  }

  // No trump, no Joker, can't win → dump.
  return strategicDump(legal, hand, trumpSuit);
}

function strategicDump(legal: string[], hand: string[], trumpSuit?: SuitName): string {
  const dumpable = legal.filter(c => c !== 'JO');
  if (dumpable.length === 0) return legal[0];

  // Tier 1: Singletons in non-trump suits (create voids).
  const singles = dumpable.filter(c => {
    const d = decodeCard(c);
    if (d.suit === trumpSuit) return false;
    return hand.filter(h => {
      const hd = decodeCard(h);
      return !hd.isJoker && hd.suit === d.suit;
    }).length === 1;
  });
  if (singles.length > 0) {
    singles.sort((a, b) => rankVal(a) - rankVal(b));
    return singles[0];
  }

  // Tier 2: Avoid breaking honor combos.
  const safe = dumpable.filter(c => {
    const d = decodeCard(c);
    if (d.suit === trumpSuit) return false;
    if (d.rank === 'king' && hand.some(h => { const hd = decodeCard(h); return hd.suit === d.suit && hd.rank === 'ace'; })) return false;
    if (d.rank === 'queen' && hand.some(h => { const hd = decodeCard(h); return hd.suit === d.suit && hd.rank === 'king'; })) return false;
    return true;
  });
  if (safe.length > 0) {
    safe.sort((a, b) => rankVal(a) - rankVal(b));
    return safe[0];
  }

  // Tier 3: Lowest non-trump.
  const nonTrump = dumpable.filter(c => decodeCard(c).suit !== trumpSuit);
  if (nonTrump.length > 0) {
    nonTrump.sort((a, b) => rankVal(a) - rankVal(b));
    return nonTrump[0];
  }

  // Only trump remains — prefer Joker over wasting a trump.
  if (legal.includes('JO')) return 'JO';

  dumpable.sort((a, b) => rankVal(a) - rankVal(b));
  return dumpable[0];
}

function rankVal(code: string): number {
  const d = decodeCard(code);
  return d.isJoker ? 15 : RANK_VALUES[d.rank!];
}

function lowest(cards: string[]): string {
  const nonJoker = cards.filter(c => c !== 'JO');
  if (nonJoker.length === 0) return cards[0];
  nonJoker.sort((a, b) => rankVal(a) - rankVal(b));
  return nonJoker[0];
}

function highest(cards: string[]): string {
  const nonJoker = cards.filter(c => c !== 'JO');
  if (nonJoker.length === 0) return cards[0];
  nonJoker.sort((a, b) => rankVal(b) - rankVal(a));
  return nonJoker[0];
}

function currentWinner(plays: TrickPlay[], trumpSuit: SuitName | undefined, ledSuit: SuitName): TrickPlay | null {
  if (plays.length === 0) return null;
  let best = plays[0];
  for (let i = 1; i < plays.length; i++) {
    if (beatsCard(plays[i].card, best.card, trumpSuit, ledSuit)) best = plays[i];
  }
  return best;
}

function cardsBeating(candidates: string[], plays: TrickPlay[], trumpSuit: SuitName | undefined, ledSuit: SuitName): string[] {
  const best = currentWinner(plays, trumpSuit, ledSuit);
  if (!best) return candidates;
  return candidates.filter(c => beatsCard(c, best.card, trumpSuit, ledSuit));
}

function leadFromLongestSuit(legal: string[], trumpSuit?: SuitName): string {
  const nonTrump = legal.filter(c => {
    const d = decodeCard(c);
    return !d.isJoker && d.suit !== trumpSuit;
  });
  if (nonTrump.length === 0) {
    const nonJoker = legal.filter(c => c !== 'JO');
    nonJoker.sort((a, b) => rankVal(b) - rankVal(a));
    return nonJoker.length > 0 ? nonJoker[0] : legal[0];
  }

  const groups = new Map<SuitName, string[]>();
  for (const c of nonTrump) {
    const suit = decodeCard(c).suit!;
    if (!groups.has(suit)) groups.set(suit, []);
    groups.get(suit)!.push(c);
  }

  // Sort by length descending, then by max rank descending (Dart tiebreak).
  const sortedSuits = [...groups.entries()].sort((a, b) => {
    const lenDiff = b[1].length - a[1].length;
    if (lenDiff !== 0) return lenDiff;
    const aMax = Math.max(...a[1].map(rankVal));
    const bMax = Math.max(...b[1].map(rankVal));
    return bMax - aMax;
  });

  const best = sortedSuits[0][1];
  best.sort((a, b) => rankVal(a) - rankVal(b));
  return best[0];
}
